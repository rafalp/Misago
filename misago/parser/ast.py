from copy import deepcopy

from django.http import HttpRequest

from .hooks import filter_ast_hook


def filter_ast(
    ast: list, content_type: str | None = None, request: HttpRequest | None = None
) -> list:
    filtered_ast = deepcopy(ast)
    filtered_ast = strip_blank_lines_ast(filtered_ast)
    return filter_ast_hook(_filter_ast_action, ast, content_type, request)


def _filter_ast_action(
    ast: list, content_type: str | None = None, request: HttpRequest | None = None
) -> list:
    return ast


def strip_blank_lines_ast(ast: list) -> list:
    clean_ast = []
    for block_ast in ast:
        if block_ast["type"] == "blank_line":
            continue
        else:
            if "children" in block_ast and block_ast["children"]:
                block_ast["children"] = strip_blank_lines_ast(block_ast["children"])
            clean_ast.append(block_ast)
    return clean_ast
