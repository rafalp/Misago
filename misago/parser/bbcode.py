from typing import Callable

from markdown_it.rules_block.state_block import StateBlock
from markdown_it.token import Token

BBCodeBlockStart = tuple[str, dict | None, int, int]
BBCodeBlockEnd = tuple[str, int, int]
BBCodeBlockStartRule = Callable[[StateBlock, int], BBCodeBlockStart | None]
BBCodeBlockEndRule = Callable[[StateBlock, int], BBCodeBlockEnd | None]


class BBCodeBlockRule:
    name: str
    element: str
    start: BBCodeBlockStartRule
    end: BBCodeBlockEndRule

    def __init__(
        self,
        name: str,
        element: str,
        start: BBCodeBlockStartRule,
        end: BBCodeBlockEndRule,
    ):
        self.name = name
        self.element = element

        self.start = start
        self.end = end

    def __call__(
        self, state: StateBlock, startLine: int, endLine: int, silent: bool
    ) -> bool:
        if state.is_code_block(startLine):
            return False

        start, end = self.scan_full_line(state, startLine)
        if start and end:
            if silent:
                return True

            return self.parse_single_line(state, startLine, start, end)

        if not start:
            return False

        return self.parse_multiple_lines(state, startLine, endLine, silent, start)

    def scan_full_line(
        self, state: StateBlock, line: int
    ) -> tuple[BBCodeBlockStart | None, BBCodeBlockEnd | None]:
        start = self.scan_line_for_start(state, line)

        if start:
            end = self.scan_line_for_end(state, line, True)
            if end and end[1] > start[3]:
                return start, end

        return start, None

    def scan_line_for_start(
        self, state: StateBlock, line: int
    ) -> BBCodeBlockStart | None:
        return self._scan_line_from_start(state, line, self.start)

    def scan_line_for_end(
        self, state: StateBlock, line: int, closing: bool = False
    ) -> BBCodeBlockEnd | None:
        if closing:
            return self._scan_line_from_end(state, line, self.end)

        return self._scan_line_from_start(state, line, self.end)

    def _scan_line_from_start(
        self,
        state: StateBlock,
        line: int,
        rule: BBCodeBlockStartRule | BBCodeBlockEndRule,
    ) -> BBCodeBlockStart | BBCodeBlockEnd | None:
        org_bmark = state.bMarks[line]
        maximum = state.eMarks[line]
        steps = 0

        try:
            while (state.bMarks[line] + state.tShift[line]) < maximum and steps < 3:
                if match := rule(state, line):
                    return match

                if state.src[state.bMarks[line] + state.tShift[line]]:
                    return None

                state.bMarks[line] += 1
                steps += 1

            return None
        except Exception:
            raise
        finally:
            state.bMarks[line] = org_bmark

    def _scan_line_from_end(
        self,
        state: StateBlock,
        line: int,
        rule: BBCodeBlockEndRule,
    ) -> BBCodeBlockEnd | None:
        org_bmark = state.bMarks[line]
        state.bMarks[line] = state.eMarks[line]
        maximum = state.eMarks[line]

        try:
            while state.bMarks[line] >= org_bmark:
                if match := rule(state, line):
                    if state.src[match[2] : maximum].strip():
                        return None

                    return match

                state.bMarks[line] -= 1

            return None
        except Exception:
            raise
        finally:
            state.bMarks[line] = org_bmark

    def parse_single_line(
        self,
        state: StateBlock,
        startLine: int,
        start: BBCodeBlockStart,
        end: BBCodeBlockEnd,
    ):
        content_start = start[3]
        content_end = end[1]
        content = state.src[content_start:content_end].strip()

        old_parent_type = state.parentType

        state.parentType = self.name
        self.state_push_open_token(state, startLine, startLine, start)

        state.parentType = "paragraph"
        token = state.push("paragraph_open", "p", 1)
        token.map = [startLine, startLine]

        token = state.push("inline", "", 0)
        token.content = content
        token.map = [startLine, state.line]
        token.children = []

        token = state.push("paragraph_close", "p", -1)

        self.state_push_close_token(state, end)

        state.parentType = old_parent_type
        state.line += 1

        return True

    def parse_multiple_lines(
        self,
        state: StateBlock,
        startLine: int,
        endLine: int,
        silent: bool,
        start: BBCodeBlockStart,
    ) -> bool:
        line = startLine
        pos = state.bMarks[line] + state.tShift[line] + start[3]
        maximum = state.eMarks[line]

        if state.src[pos:maximum].strip():
            return False

        end = None
        nesting = 1

        while line < endLine:
            line += 1

            if (
                state.isEmpty(line)
                or state.is_code_block(line)
                or all(self.scan_full_line(state, line))
                or line > state.lineMax
            ):
                continue

            if self.scan_line_for_start(state, line):
                nesting += 1

            elif match := self.scan_line_for_end(state, line):
                nesting -= 1

                if nesting == 0:
                    end = match
                    break

        if silent or not end:
            return nesting == 0

        self.state_push_open_token(state, startLine, line, start)
        self.state_push_children(state, startLine + 1, line)
        self.state_push_close_token(state, end)

        state.line = line + 1
        return True

    def state_push_open_token(
        self, state: StateBlock, startLine: int, endLine: int, start: BBCodeBlockStart
    ) -> Token:
        token = state.push(f"{self.name}_open", self.element, 1)
        token.markup = start[0]
        token.map = [startLine, endLine]

        if attrs := start[1]:
            for attr_name, attr_value in attrs.items():
                token.attrSet(attr_name, attr_value)

            if meta := self.get_meta(attrs):
                token.meta = meta

        return token

    def state_push_close_token(self, state: StateBlock, end: BBCodeBlockEnd) -> Token:
        token = state.push(f"{self.name}_close", self.element, -1)
        token.markup = end[0]
        return token

    def state_push_void_token(
        self, state: StateBlock, startLine: int, start: BBCodeBlockStart
    ) -> Token:
        token = state.push(f"{self.name}", self.element, 0)
        token.markup = start[0]
        token.map = [startLine, startLine]

        if attrs := start[1]:
            for attr_name, attr_value in attrs.items():
                token.attrSet(attr_name, attr_value)

            if meta := self.get_meta(attrs):
                token.meta = meta

        return token

    def state_push_children(self, state: StateBlock, startLine: int, endLine: int):
        if startLine == endLine:
            return

        old_line_max = state.lineMax
        old_parent = state.parentType

        state.lineMax = startLine
        state.parentType = self.name

        state.level += 1
        state.md.block.tokenize(state, startLine, endLine)
        state.level -= 1

        state.lineMax = old_line_max
        state.parentType = old_parent

    def get_meta(self, attrs: dict) -> dict | None:
        return None


def bbcode_block_start_rule(
    bbcode: str, state: StateBlock, line: int, args: bool = False
) -> tuple[str, str | None, int, int] | None:
    start = state.bMarks[line] + state.tShift[line]
    maximum = state.eMarks[line]
    src = state.src[start:maximum]

    block_bbcode = f"[{bbcode}".lower()
    block_bbcode_len = len(block_bbcode)

    if src.lower()[:block_bbcode_len] != block_bbcode:
        return None

    if "]" not in src[block_bbcode_len:]:
        return None

    end = src.index("]", 0, maximum - start)
    if end == block_bbcode_len:
        return src[: block_bbcode_len + 1], None, start, start + end + 1

    if src[block_bbcode_len] != "=" or not args:
        return None

    args_str = src[block_bbcode_len + 1 : end]
    if args_str and (
        (args_str[0] == '"' and args_str[-1] == '"')
        or (args_str[0] == "'" and args_str[-1] == "'")
    ):
        args_str = args_str[1:-1]

    args_str = args_str.strip() or None

    return src[: end + 1], args_str, start, end + 1


def bbcode_block_end_rule(
    bbcode: str, state: StateBlock, line: int
) -> tuple[str, int, int] | None:
    start = state.bMarks[line] + state.tShift[line]
    maximum = state.eMarks[line]
    src = state.src[start:maximum]

    block_bbcode = f"[/{bbcode}]".lower()
    block_bbcode_len = len(block_bbcode)

    if src[:block_bbcode_len].lower() == block_bbcode:
        return src[:block_bbcode_len], start, start + block_bbcode_len

    return None
