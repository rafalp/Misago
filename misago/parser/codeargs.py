import re

from markdown_it.common.utils import unescapeAll

from .pygments import PYGMENTS_LANGUAGES


CODE_ARGS = re.compile(r"^(?P<info>.+)[;,] *syntax[:=] *(?P<syntax>.+) *$")


def parse_code_args(args_str: str) -> dict | None:
    if args_str.lower().startswith("syntax:") or args_str.lower().startswith("syntax="):
        syntax = unescapeAll(args_str[7:].strip())
        if not syntax:
            return None
        return {"syntax": syntax.lower()}

    if match := CODE_ARGS.match(args_str):
        args = {}

        if info := match.group("info").strip():
            args["info"] = unescapeAll(info)
        if syntax := match.group("syntax").strip():
            args["syntax"] = unescapeAll(syntax)

        return args if args else None

    if args_str.lower() in PYGMENTS_LANGUAGES:
        return {"syntax": args_str.lower()}

    if args_str:
        return {"info": unescapeAll(args_str)}

    return None
