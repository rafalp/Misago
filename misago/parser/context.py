from dataclasses import dataclass

from django.contrib.auth import get_user_model
from django.contrib.auth.models import AnonymousUser
from django.http import HttpRequest

from ..conf.dynamicsettings import DynamicSettings
from ..permissions.proxy import UserPermissionsProxy
from .forumaddress import ForumAddress
from .hooks import setup_parser_context_hook

User = get_user_model()


@dataclass(frozen=True)
class ParserContext:
    content_type: str | None
    forum_address: ForumAddress
    request: HttpRequest | None
    user: User | AnonymousUser
    user_permissions: UserPermissionsProxy
    cache_versions: dict
    settings: DynamicSettings
    plugin_data: dict


def create_parser_context(
    request: HttpRequest | None = None,
    *,
    user_permissions: UserPermissionsProxy | None = None,
    cache_versions: dict | None = None,
    settings: DynamicSettings | None = None,
    content_type: str | None = None,
) -> ParserContext:
    data = {
        "user_permissions": None,
        "user": None,
        "cache_versions": None,
        "settings": None,
        "plugin_data": {},
    }

    if user_permissions:
        data["user"] = user_permissions.user
        data["user_permissions"] = user_permissions
    elif request:
        data["user"] = request.user
        data["user_permissions"] = request.user_permissions
    else:
        raise TypeError(
            "'user_permissions' argument is required if 'request' is not set"
        )

    if cache_versions:
        data["cache_versions"] = cache_versions
    elif request:
        data["cache_versions"] = request.cache_versions
    else:
        raise TypeError("'cache_versions' argument is required if 'request' is not set")

    if settings:
        data["settings"] = settings
    elif request:
        data["settings"] = request.settings
    else:
        raise TypeError("'settings' argument is required if 'request' is not set")

    context = ParserContext(
        content_type=content_type,
        forum_address=ForumAddress(data["settings"]),
        request=request,
        **data,
    )

    if not setup_parser_context_hook:
        return context

    return setup_parser_context_hook(_context_lambda, context)


_context_lambda = lambda c: c
