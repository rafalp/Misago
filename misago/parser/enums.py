from enum import StrEnum


class ContentType(StrEnum):
    GROUP_DESCRIPTION = "group-description"
    POST = "post"
    SIGNATURE = "signature"
