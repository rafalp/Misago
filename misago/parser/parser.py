import re
from functools import cached_property
from typing import Callable

from django.utils.crypto import get_random_string


class Pattern:
    pattern_type: str
    pattern: str

    def parse(
        self, parser: "Parser", match: str, parents: list[str]
    ) -> dict | list[dict]:
        raise NotImplementedError()


class LineBreak(Pattern):
    pattern_type = "line-break"
    pattern = r" *\n *"

    def parse(self, parser: "Parser", match: str, parents: list[str]) -> dict:
        return {"type": self.pattern_type}


class Parser:
    block_patterns: list[Pattern]
    inline_patterns: list[Pattern]
    post_processors: list[Callable[["Parser", list[dict]], list[dict]]]

    reserve_inline_code = re.compile(r"`*`(.|\n)+?``*")
    _reserved_patterns: dict[str, str]

    def __init__(
        self,
        block_patterns: list[Pattern] | None = None,
        inline_patterns: list[Pattern] | None = None,
        post_processors: list[Callable[["Parser", list[dict]], list[dict]]]
        | None = None,
    ):
        self.block_patterns = block_patterns or []
        self.inline_patterns = inline_patterns or []
        self.post_processors = post_processors or []

        self._reserved_patterns = {}

    def __call__(self, markup: str) -> list[dict]:
        markup = self.reserve_patterns(markup)
        ast = self.parse_blocks(markup, [])
        for post_processor in self.post_processors:
            ast = post_processor(self, ast)
        return ast

    def reserve_patterns(self, markup: str) -> str:
        if not "`" in markup:
            return markup

        def replace_pattern(match):
            match_str = match.group(0)
            if match_str.startswith("``") or match_str.endswith("``"):
                return match_str

            pattern_id = f"%%{get_random_string(12)}%%"
            while pattern_id in markup or pattern_id in self._reserved_patterns:
                pattern_id = f"%%{get_random_string(12)}%%"

            self._reserved_patterns[pattern_id] = match_str
            return pattern_id

        return self.reserve_inline_code.sub(replace_pattern, markup)

    def reverse_reservations(self, value: str) -> str:
        if not self._reserved_patterns or "%%" not in value:
            return value

        for pattern, org in self._reserved_patterns.items():
            value = value.replace(pattern, org)
        return value

    def parse_blocks(self, markup: str, parents: list[str]) -> list[dict]:
        cursor = 0

        result: list[dict] = []
        for m in self._block_re.finditer(markup):
            for key, pattern in self._final_block_patterns.items():
                block_match = m.group(key)
                if block_match is not None:
                    start = m.start()
                    if start > cursor:
                        result += self.parse_paragraphs(markup[cursor:start], parents)

                    block_ast = pattern.parse(self, block_match, parents)
                    if isinstance(block_ast, list):
                        result += block_ast
                    elif isinstance(block_ast, dict):
                        result.append(block_ast)

                    cursor = m.end()
                    break

        if cursor < len(markup):
            result += self.parse_paragraphs(markup[cursor:], parents)

        return result

    def parse_paragraphs(self, markup: str, parents: list[str]) -> list[dict]:
        markup = markup.strip()

        if not markup:
            return []

        parents = parents + ["paragraph"]

        result: list[dict] = []
        for m in self._paragraph_re.finditer(markup):
            result.append(
                {
                    "type": "paragraph",
                    "children": self.parse_inline(
                        m.group(0).strip(), parents, reverse_reservations=True
                    ),
                }
            )
        return result

    def parse_inline(
        self,
        markup: str,
        parents: list[str],
        reverse_reservations: bool = False,
    ) -> list[dict]:
        if reverse_reservations:
            markup = self.reverse_reservations(markup)

        cursor = 0

        result: list[dict] = []
        for m in self._inline_re.finditer(markup):
            for key, pattern in self._final_inline_patterns.items():
                block_match = m.group(key)
                if block_match is not None:
                    start = m.start()
                    if start > cursor:
                        if result and result[-1]["type"] == "text":
                            result[-1]["text"] += markup[cursor:start]
                        else:
                            result.append(
                                {"type": "text", "text": markup[cursor:start]}
                            )

                    inline_ast = pattern.parse(self, block_match, parents)
                    if isinstance(inline_ast, list):
                        for child in inline_ast:
                            if (
                                result
                                and child["type"] == "text"
                                and result[-1]["type"] == "text"
                            ):
                                result[-1]["text"] += child["text"]
                            else:
                                result.append(child)

                    elif isinstance(inline_ast, dict):
                        if (
                            result
                            and inline_ast["type"] == "text"
                            and result[-1]["type"] == "text"
                        ):
                            result[-1]["text"] += inline_ast["text"]
                        else:
                            result.append(inline_ast)

                    cursor = m.end()
                    break

        if cursor < len(markup):
            if result and result[-1]["type"] == "text":
                result[-1]["text"] += markup[cursor:]
            else:
                result.append({"type": "text", "text": markup[cursor:]})

        return result

    @cached_property
    def _final_block_patterns(self) -> dict[str, Pattern]:
        patterns: list[Pattern] = self.block_patterns.copy()
        return {f"b_{i}": pattern for i, pattern in enumerate(patterns)}

    @cached_property
    def _block_re(self) -> re.Pattern:
        return self._build_re_pattern(self._final_block_patterns)

    @cached_property
    def _paragraph_re(self) -> re.Pattern:
        return re.compile(r".+(\n.+)*")

    @cached_property
    def _final_inline_patterns(self) -> dict[str, Pattern]:
        patterns: list[Pattern] = self.inline_patterns.copy()
        patterns += [LineBreak()]
        return {f"i_{i}": pattern for i, pattern in enumerate(patterns)}

    @cached_property
    def _inline_re(self) -> re.Pattern:
        return self._build_re_pattern(self._final_inline_patterns)

    def _build_re_pattern(self, patterns: dict[str, Pattern]) -> re.Pattern:
        return re.compile(
            "|".join(
                f"(?P<{key}>{pattern.pattern})" for key, pattern in patterns.items()
            ),
            re.IGNORECASE,
        )
