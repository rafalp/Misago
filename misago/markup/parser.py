import markdown
from markdown.extensions.fenced_code import FencedCodeExtension

from .bbcode.code import CodeBlockExtension
from .bbcode.hr import BBCodeHRProcessor
from .bbcode.inline import bold, image, italics, underline, url
from .bbcode.quote import QuoteExtension
from .bbcode.spoiler import SpoilerExtension
from .htmlparser import parse_html_string, print_html_string
from .links import clean_links, linkify_texts
from .md.shortimgs import ShortImagesExtension
from .md.strikethrough import StrikethroughExtension
from .mentions import add_mentions
from .pipeline import pipeline


def parse(
    text,
    request,
    poster,
    allow_mentions=True,
    allow_links=True,
    allow_images=True,
    allow_blocks=True,
    force_shva=False,
):
    """
    Message parser

    Utility for flavours to call

    Breaks text into paragraphs, supports code, spoiler and quote blocks,
    headers, lists, images, spoilers, text styles

    Returns dict object
    """
    md = md_factory(
        allow_links=allow_links, allow_images=allow_images, allow_blocks=allow_blocks
    )

    parsing_result = {
        "original_text": text,
        "parsed_text": "",
        "markdown": md,
        "mentions": [],
        "images": [],
        "internal_links": [],
        "outgoing_links": [],
    }

    # Parse text
    parsed_text = md.convert(text)

    # Clean and store parsed text
    parsing_result["parsed_text"] = parsed_text.strip()

    # Run additional operations
    if allow_mentions or allow_links or allow_images:
        root_node = parse_html_string(parsing_result["parsed_text"])

        if allow_links:
            linkify_texts(root_node)

        if allow_mentions:
            add_mentions(parsing_result, root_node)

        if allow_links or allow_images:
            clean_links(request, parsing_result, root_node, force_shva)

        parsing_result["parsed_text"] = print_html_string(root_node)

    # Let plugins do their magic
    parsing_result = pipeline.process_result(parsing_result)

    return parsing_result


def md_factory(allow_links=True, allow_images=True, allow_blocks=True):
    """creates and configures markdown object"""
    md = markdown.Markdown(extensions=["markdown.extensions.nl2br"])

    # Remove HTML allowances
    md.preprocessors.deregister("html_block")
    md.inlinePatterns.deregister("html")

    # Remove references
    md.parser.blockprocessors.deregister("reference")
    md.inlinePatterns.deregister("reference")
    md.inlinePatterns.deregister("image_reference")
    md.inlinePatterns.deregister("short_reference")

    # Add [b], [i], [u]
    md.inlinePatterns.register(bold, "bb_b", 55)
    md.inlinePatterns.register(italics, "bb_i", 55)
    md.inlinePatterns.register(underline, "bb_u", 55)

    # Add ~~deleted~~
    strikethrough_md = StrikethroughExtension()
    strikethrough_md.extendMarkdown(md)

    if allow_links:
        # Add [url]
        md.inlinePatterns.register(url(md), "bb_url", 155)
    else:
        # Remove links
        md.inlinePatterns.deregister("link")
        md.inlinePatterns.deregister("autolink")
        md.inlinePatterns.deregister("automail")

    if allow_images:
        # Add [img]
        md.inlinePatterns.register(image(md), "bb_img", 145)
        short_images_md = ShortImagesExtension()
        short_images_md.extendMarkdown(md)
    else:
        # Remove images
        md.inlinePatterns.deregister("image_link")

    if allow_blocks:
        # Add [hr] and [quote] blocks
        md.parser.blockprocessors.register(BBCodeHRProcessor(md.parser), "bb_hr", 45)

        fenced_code = FencedCodeExtension(lang_prefix="language-")
        fenced_code.extendMarkdown(md)

        code_bbcode = CodeBlockExtension()
        code_bbcode.extendMarkdown(md)

        quote_bbcode = QuoteExtension()
        quote_bbcode.extendMarkdown(md)

        spoiler_bbcode = SpoilerExtension()
        spoiler_bbcode.extendMarkdown(md)
    else:
        # Remove blocks
        md.parser.blockprocessors.deregister("hashheader")
        md.parser.blockprocessors.deregister("setextheader")
        md.parser.blockprocessors.deregister("code")
        md.parser.blockprocessors.deregister("quote")
        md.parser.blockprocessors.deregister("hr")
        md.parser.blockprocessors.deregister("olist")
        md.parser.blockprocessors.deregister("ulist")

    return pipeline.extend_markdown(md)
