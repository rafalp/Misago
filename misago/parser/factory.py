from typing import Callable

from django.contrib.auth import get_user_model

from .context import ParserContext
from .hooks import create_parser_hook
from .parser import Parser, Pattern
from .patterns import block_patterns, inline_patterns
from .postprocessors import post_processors

User = get_user_model()


def create_parser(context: ParserContext) -> Parser:
    return create_parser_hook(
        _create_parser_action,
        context,
        block_patterns=block_patterns.copy(),
        inline_patterns=inline_patterns.copy(),
        post_processors=post_processors.copy(),
    )


def _create_parser_action(
    context: ParserContext,
    *,
    block_patterns: list[Pattern],
    inline_patterns: list[Pattern],
    post_processors: list[Callable[[Parser, list[dict]], list[dict]]],
) -> Parser:
    return Parser(block_patterns, inline_patterns, post_processors)
