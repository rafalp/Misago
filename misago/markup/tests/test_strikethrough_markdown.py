from ..parser import parse


def test_strikethrough_markdown(request_mock, user, snapshot):
    text = "Lorem ~~ipsum~~ dolor met!"
    result = parse(text, request_mock, user)
    assert snapshot == result["parsed_text"]
