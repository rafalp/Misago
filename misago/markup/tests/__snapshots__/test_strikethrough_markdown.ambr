# serializer version: 1
# name: test_strikethrough_markdown
  '<p>Lorem <del>ipsum</del> dolor met!</p>'
# ---
